tests/test_data.py

```diff
@@ -0,0 +1,193 @@
+    obs = np.array([obs_space.sample() for _ in range(length + 1)])
+    acts = np.array([act_space.sample() for _ in range(length)])
+    infos = [{} for _ in range(length)]
+    rews = np.random.randn(length)
+
+    for cls in [data.Trajectory, functools.partial(data.TrajectoryWithRew, rews=rews)]:
+        with pytest.raises(
+            ValueError, match=r"expected one more observations than actions.*"
+        ):
+            cls(obs=obs[:-1], acts=acts, infos=None)
+        with pytest.raises(
+            ValueError, match=r"expected one more observations than actions.*"
+        ):
+            cls(obs=obs, acts=acts[:-1], infos=None)
+
+        with pytest.raises(
+            ValueError, match=r"infos when present must be present for each action.*"
+        ):
+            cls(obs=obs, acts=acts, infos=infos[:-1])
+        with pytest.raises(
+            ValueError, match=r"infos when present must be present for each action.*"
+        ):
+            cls(obs=obs[:-1], acts=acts[:-1], infos=infos)
+
+    _check_1d_shape(
+        fn=lambda rews: data.TrajectoryWithRew(
+            obs=obs, acts=acts, infos=infos, rews=rews
+        ),
+        length=length,
+        expected_msg=r"rewards must be 1D array.*",
+    )
+
+    with pytest.raises(ValueError, match=r"rewards dtype.* not a float"):
+        data.TrajectoryWithRew(
+            obs=obs, acts=acts, infos=infos, rews=np.zeros(length, dtype=np.int)
+        )
+
+
+@pytest.mark.parametrize("obs_space", OBS_SPACES)
+@pytest.mark.parametrize("act_space", ACT_SPACES)
+@pytest.mark.parametrize("length", [0, 1, 2, 10])
+def test_valid_transitions(
+    obs_space: gym.Space, act_space: gym.Space, length: int
+) -> None:
+    """Checks trajectories can be created for a variety of lengths and spaces."""
+    obs = np.array([obs_space.sample() for _ in range(length)])
+    next_obs = np.array([obs_space.sample() for _ in range(length)])
+    acts = np.array([act_space.sample() for _ in range(length)])
+    dones = np.zeros(length, dtype=np.bool)
+    rews = np.random.randn(length)
+
+    trans = data.Transitions(obs=obs, acts=acts, next_obs=next_obs, dones=dones)
+    trans_rew = data.TransitionsWithRew(
+        obs=obs, acts=acts, next_obs=next_obs, dones=dones, rews=rews
+    )
+    for traj in [trans, trans_rew]:
+        assert len(traj) == length
+
+
+@pytest.mark.parametrize("obs_space", OBS_SPACES)
+@pytest.mark.parametrize("act_space", ACT_SPACES)
+@pytest.mark.parametrize("length", [1, 2, 10])
+def test_invalid_transitions(
+    obs_space: gym.Space, act_space: gym.Space, length: int
+) -> None:
+    """Checks input validation catches space and dtype related errors."""
+    obs = np.array([obs_space.sample() for _ in range(length)])
+    next_obs = np.array([obs_space.sample() for _ in range(length)])
+    acts = np.array([act_space.sample() for _ in range(length)])
+    dones = np.zeros(length, dtype=np.bool)
+    rews = np.random.randn(length)
```

Comment:
In fact, since every test here parameterizes over the same `obs_space` and `act_space`. It probably would be good to put everyone in a `TestData` class with fixtures/attributes for transitions and trajectories.